Index: src/couchdb/priv/Makefile.am
===================================================================
--- src/couchdb/priv/Makefile.am	(revision 1129890)
+++ src/couchdb/priv/Makefile.am	(working copy)
@@ -78,9 +78,9 @@
 	        $(LN_S) couch_icu_driver couch_icu_driver.so; \
 	fi
 if WINDOWS
-	$(INSTALL) $(ICU_LOCAL_BIN)/icuuc42.dll $(bindir)
-	$(INSTALL) $(ICU_LOCAL_BIN)/icudt42.dll $(bindir)
-	$(INSTALL) $(ICU_LOCAL_BIN)/icuin42.dll $(bindir)
+	$(INSTALL) $(ICU_LOCAL_BIN)/icuuc44.dll $(bindir)
+	$(INSTALL) $(ICU_LOCAL_BIN)/icudt44.dll $(bindir)
+	$(INSTALL) $(ICU_LOCAL_BIN)/icuin44.dll $(bindir)
 	$(INSTALL) $(JS_LIB_BINARY) $(bindir)
 	$(INSTALL) .libs/couchspawnkillable.exe \
 		"$(DESTDIR)$(couchprivdir)/couchspawnkillable.exe"
Index: INSTALL.Windows
===================================================================
--- INSTALL.Windows	(revision 1129890)
+++ INSTALL.Windows	(working copy)
@@ -9,7 +9,7 @@
 You will need the following installed:
 
  * Erlang OTP (=14B01)        (http://erlang.org/)
- * ICU                        (http://icu.sourceforge.net/)
+ * ICU        (=4.4.*)         (http://icu.sourceforge.net/)
  * OpenSSL                    (http://www.openssl.org/)
  * Mozilla SpiderMonkey (1.8) (http://www.mozilla.org/js/spidermonkey/)
  * libcurl                    (http://curl.haxx.se/libcurl/)
@@ -21,7 +21,7 @@
 
  * When installing Erlang, you must build it from source.
 
-   The CouchDB build makes use of a number of the Erlang build scripts.
+The CouchDB build requires a number of the Erlang build scripts.
 
  * When installing ICU, select the binaries built with Visual Studio 2008.
 
