diff --git a/erts/etc/win32/nsis/find_redist.sh b/erts/etc/win32/nsis/find_redist.sh
index 153977d..c2057f8 100755
--- a/erts/etc/win32/nsis/find_redist.sh
+++ b/erts/etc/win32/nsis/find_redist.sh
@@ -134,7 +134,7 @@ fi
 
 #echo $BPATH
 for BP in $BPATH_LIST; do
-    for verdir in "sdk v2.0" "sdk v3.5" "v6.0A"; do
+    for verdir in "sdk v2.0" "sdk v3.5" "v6.0A" "v7.0" "v7.1"; do
 	BPATH=$BP
 	fail=false
 	for x in $verdir bootstrapper packages vcredist_x86 vcredist_x86.exe; do
@@ -157,6 +157,18 @@ for BP in $BPATH_LIST; do
     fi
 done
 
+# shortcut for locating vcredist_x86.exe is to put it into $ERL_TOP
+if [ -f $ERL_TOP/vcredist_x86.exe ]; then
+    echo $ERL_TOP/vcredist_x86.exe
+    exit 0
+fi
+
+# or $ERL_TOP/.. to share across multiple builds
+if [ -f $ERL_TOP/../vcredist_x86.exe ]; then
+    echo $ERL_TOP/../vcredist_x86.exe
+    exit 0
+fi
+
 echo "Failed to locate vcredist_x86.exe because directory structure was unexpected" >&2
 exit 3
 

